### PR TITLE
Collection Instrument page redesign

### PR DESCRIPTION
**Pull request by SteveScorfield**
# What and why?
As part of the rOps redesign, a new endpoint was required that would allow for a user to add and remove CIs in one go via the rOps CI page. This endpoint deals with both adding and removing CIs in one database transaction. Any failures will roll back the changes and the offending action (add or remove) will be reported to the user. 

# How to test?
Add and remove CIs in rOps. 
Run the unit tests
Run acceptance tests

Trello
https://trello.com/c/ELDnXWO2/2344-rops-design-improvements-ci-page-redesign

Jira
https://jira.ons.gov.uk/browse/RAS-266

### CODE REVIEW

**Comment by SteveScorfield** [comment]
[rOps design improvements - CI page redesign](https://trello.com/c/ELDnXWO2/2344-rops-design-improvements-ci-page-redesign)

**Review by LJBabbage — CHANGES_REQUESTED** [review]
The code works, but it needs a bit of refactoring and refinement, coverage is a little short, although the changes below might sort that out. We might want to consider having a test for a db failure rather than publish one

File: application/controllers/collection_instrument.py [hunk 1]
```diff
@@ -239,6 +238,73 @@ def upload_instrument_with_no_collection_exercise(self, survey_id, classifiers=N
 
         return instrument
 
+    @with_db_session
+    def multi_instrument_selection_to_exercise(self, instrument_selection, exercise_id, session=None):
```
[the remark this answers is not in the document]
Comment (SteveScorfield) [reply application/controllers/collection_instrument.py:242]: That makes sense. I've updated this.

File: application/controllers/collection_instrument.py [hunk 2]
```diff
@@ -239,6 +238,73 @@ def upload_instrument_with_no_collection_exercise(self, survey_id, classifiers=N
...
+        if linked_instruments:
+            for i in linked_instruments.instruments:
+                current_instruments.append(str(i.instrument_id))
+
+            selected_instruments = set(sorted(instrument_selection)).difference(sorted(current_instruments))
+            unselected_instruments = set(sorted(current_instruments)).difference(sorted(instrument_selection))
+        else:
+            selected_instruments = set(instrument_selection)
+
+        validate_uuid(exercise_id)
```
[the remark this answers is not in the document]
Comment (SteveScorfield) [reply application/controllers/collection_instrument.py:265]: That's been moved

File: application/controllers/collection_instrument.py [hunk 3]
```diff
@@ -239,6 +238,73 @@ def upload_instrument_with_no_collection_exercise(self, survey_id, classifiers=N
...
+        Link a collection instrument to a collection exercise
+
+        :param instrument_selection: A list of instrument ids
+        :param exercise_id: A collection exercise id (UUID)
+        :param session: database session
+        :return: True if instrument has been successfully linked to exercise
+        """
+        current_instruments = []
+        unselected_instruments = {}
+        linked_instruments = query_exercise_by_id(exercise_id, session)
```
[the remark this answers is not in the document]
Comment (SteveScorfield) [reply application/controllers/collection_instrument.py:253]: It returns everything related to that exercise, including all the CIs. By using this, we don't need to call any other queries to get more info.

[the remark this answers is not in the document]
Comment (LJBabbage) [reply application/controllers/collection_instrument.py:253]: It's the name I was suggesting should be exercise as that is the return on the query

File: application/controllers/collection_instrument.py [hunk 4]
```diff
@@ -267,6 +266,52 @@ def upload_eq(self, survey_id, classifiers=None, session=None):
...
+        :param instrument_selection: A list of instrument ids
+        :param exercise_id: A collection exercise id (UUID)
+        :param session: database session
+        :return: True if instrument has been successfully linked to exercise
+        """
+        validate_uuid(exercise_id)
+
+        linked_instruments = []
+
+        # This query will eitheer create a CE (with no CIs attached) or will pull in all linked CIs for the exercise id
```
Comment (LJBabbage) [inline application/controllers/collection_instrument.py:283]: Minor typo

File: application/controllers/collection_instrument.py [hunk 5]
```diff
@@ -267,6 +266,52 @@ def upload_eq(self, survey_id, classifiers=None, session=None):
         session.add(instrument)
         return instrument
 
+    @with_db_session
+    def update_collection_exercise_instruments(self, instrument_selection, exercise_id, session=None):
+        """
+        Link a collection instrument to a collection exercise
```
Comment (LJBabbage) [inline application/controllers/collection_instrument.py:272]: Just needs a tweak in description

**Review by LJBabbage — CHANGES_REQUESTED** [review]
Looking much better, a slightly strange test is the only real thing I can see

**Pull request merged**